Maple2.Server.Game/Manager/ItemBoxManager.cs

```diff
@@ -0,0 +1,259 @@
+        }
+
+        IndividualItemDropTable.Entry selectedEntry = drops[index];
+        for (int startCount = 0; startCount < count; startCount++) {
+            if (!session.Item.Inventory.ConsumeItemComponents(new[] {
+                    new ItemComponent(item.Id, -1, itemRequiredAmount, ItemTag.None),
+                })) {
+                return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+            }
+
+            IEnumerable<Item> itemList = session.ItemBox.GetItemsFromGroup(selectedEntry);
+            foreach (Item newItem in itemList) {
+                if (!session.Item.Inventory.Add(newItem, true, true)) {
+                    error = ItemBoxError.s_err_cannot_open_multi_itembox_inventory;
+                }
+            }
+            BoxCount++;
+            if (error != ItemBoxError.ok) {
+                return error;
+            }
+        }
+        return error;
+    }
+
+    private ItemBoxError OpenItemBox(Item item, int globalDropBoxId, int unknownId, int itemId, int boxId, int itemRequiredAmount, int count = 1) {
+        var error = ItemBoxError.ok;
+        // if globalDropBoxId > 0, unknownId is possibly another globalDropBoxId. Not confirmed, just a guess.
+        // otherwise unknownId is treated as a bool. if 1 = receive one item from each drop group, if 0 = receive all items from one drop group
+
+        if (boxId == 0) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        if (itemRequiredAmount > -1 && session.Item.Inventory.Find(item.Id).Sum(box => box.Amount) < itemRequiredAmount) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        // TODO: branch off to globalDropBoxId. Temp solution
+        if (globalDropBoxId > 0) {
+            return HandleGlobalDropBox(item, itemId, itemRequiredAmount, count);
+        }
+
+        // Get dropbox
+        if (!session.TableMetadata.IndividualItemDropTable.Entries.TryGetValue(boxId, out Dictionary<byte, IList<IndividualItemDropTable.Entry>>? dropGroupTable)) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        if (unknownId == 1) {
+            for (int startCount = 0; startCount < count; startCount++) {
+                if (!session.Item.Inventory.ConsumeItemComponents(new[] {
+                        new ItemComponent(item.Id, item.Rarity, itemRequiredAmount, ItemTag.None),
+                    })) {
+                    return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+                }
+
+                if (itemId > 0 && session.ItemMetadata.TryGet(itemId, out ItemMetadata? itemMetadata)) {
+                    if (!session.Item.Inventory.Add(new Item(itemMetadata, item.Metadata.Option?.ConstantId ?? 1), true, true)) {
+                        error = ItemBoxError.s_err_cannot_open_multi_itembox_inventory;
```

Comment:
just to double check, we intentionally continue when we fail here?

Reply:
Yeah the entire box opening process needs to finish before failing/stopping